### PR TITLE
DS-4316: Indirect entity refs during csv import

### PR DESCRIPTION
**Pull request by cwilper**
https://jira.duraspace.org/browse/DS-4316

This adds two new ways (in addition to the existing item UUID) to refer to other items when adding relationships via CSV import:

A) By row name. If a previous row in the same CSV declares a name (via an optional `rowName` column), you may refer to it by whatever name it has been given in the CSV, by specifying `rowName:NAME` in the relationship.label column (where NAME is the actual unique row name of the row you're referring to). The row name may be any unique string. It is not persisted as metadata; the name is just a convenience used during import.

B) By unique metadata value. If a previous row in the same CSV, or an item already in the repository, has a unique value for a given metadata field, you may refer to that item by specifying `FIELD:VALUE` in the relationship.label column. This allows, for example, referring to another item by DOI (dc.identifier.doi:SOME_DOI) or another application-specific identifier you know to be unique.

The CSV import will fail with an appropriate error message if:

* Any reference doesn't refer to an existing row or item
* Any reference resolves to MULTIPLE items

Note:

The MetadataImport script has historically behaved slightly differently depending on whether you pass in -s ("silent") or not. In the former case, it does not run the initial pass through the CSV (the validation/verification part). Since it is important to test the first pass-through, we have added a -v ("validate") option to the CLI, that ONLY runs the initial validation. This is used by a limited set of the tests to ensure failure cases behave properly.

TBD: Documentation needs to be added for this capability (CLI usage and MetadataImport docs in general require updates)

### CODE REVIEW

**Comment by tdonohue** [comment]
@cwilper : I like the concepts/ideas behind 2 (`by row name`) and 3 (`by unique metadata value`) in your description...those seem like great additions to me.  

However, I'm worried about 1 (`by row number`), as I'm worried it could invite a lot of accidental human error.  Row numbers are complex because you'd need to be very clear whether the heading row is row 1, or if the first non-heading row is row 1.  If the heading row is not counted, this further complicates things, as the row numbers in Excel will show the heading row as row 1 (and as humans, we might assume the Excel row numbers are accurate & want to use them).   If you see what I mean, row numbers could be prone to human error, and you may end up accidentally linking the wrong items together.

All that said though, I think that 2 and 3 are "good enough" solutions here.  Perhaps we should simply consider dropping the `by row number` option, as it's potentially more prone to human error.

I'll note, I haven't reviewed the code here yet (as you noted it as a Proof of Concept). Just giving feedback on the ideas.

**Comment by tdonohue** [comment]
One final note,  if we are calling option 2 `by row name`, then the column used should be named `rowName` (not `rowId`).  I think the `rowId` column is problematic anyways as it overlaps in meanings with the `id` column.  E.g. if you were to say "row id", that could either refer to a row's `id` column, or a "rowId" column.

**Comment by cwilper** [comment]
Thanks for the feedback @tdonohue . We have rebased this work on latest master, applied the suggested changes (omitting the implicit row# style reference, and renaming rowId to rowName), and added several tests.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/MetadataValueDAOImpl.java [hunk 1]
```diff
@@ -49,6 +49,17 @@ protected MetadataValueDAOImpl() {
         return list(context, criteriaQuery, false, MetadataValue.class, -1, -1);
     }
 
+    @Override
+    public Iterator<MetadataValue> findByFieldAndValue(Context context, MetadataField metadataField, String value)
+            throws SQLException {
+        String queryString = "SELECT m from MetadataValue m where metadata_field_id = :metadata_field_id " +
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/MetadataValueDAOImpl.java]: Is the DB optimized for this type of queries?
Should this be limited to a specific DSO type (e.g. only items)?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+     * a metadata field and value reference with multiable references
+     */
+    @Test
+    public void testMultiMdRef() throws Exception {
+        String[] csv = {"id,relationship.type,relation.isAuthorOfPublication,collection,dc.identifier.other",
+        "+,Person,," +  col1.getHandle() + ",0",
+        "+,Person,," +  col1.getHandle() + ",1",
+        "+,Publication,dc.identifier.other:0||dc.identifier.other:1," +  col1.getHandle() + ",2"};
+        Item[] items = runImport(csv);
+        assertRelationship(items[2], items[0], 1);
```
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Can you also verify the order matches

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    @Test
+    public void testMultiRefArchivedCsv() throws Exception {
+        Item person = ItemBuilder.createItem(context, col1)
+                .withTitle("Person")
+                .withRelationshipType("Person")
+                .build();
+        String[] csv = {"id,dc.title,relationship.type,relation.isAuthorOfPublication,collection,rowName," +
+                "dc.identifier.other",
+                "+,Person2,Person,," +  col1.getHandle() + ",idVal,0",
+                "+,Pub1,Publication,dc.title:Person||dc.title:Person2," +  col1.getHandle() + ",anything,1"};
```
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Does this also support mixing a metadata, UUID, rowName lookup, and still retain the order?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+    /**
+     * Test existence of newly created items with proper relationships defined in mets via
+     * a metadata field and value reference
+     */
+    @Test
+    public void testSingleMdRef() throws Exception {
+        String[] csv = {"id,relationship.type,relation.isAuthorOfPublication,collection,dc.identifier.other",
+                "+,Person,," +  col1.getHandle() + ",0",
+                "+,Publication,dc.identifier.other:0," +  col1.getHandle() + ",1"};
```
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: can we also test this will fail correctly if the referenced person occurs in the CSV after the publication

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+    /**
+     * Test failure when referring to item by non unique metadata in the csv file.
+     */
+    @Test
+    public void testNonUniqueMDRefInCsv() throws Exception {
+        String[] csv = {"id,relationship.type,relation.isAuthorOfPublication,collection,dc.identifier.other",
+                "+,Person,," + col1.getHandle() + ",1",
+                "+,Person,," + col1.getHandle() + ",1",
+                "+,Publication,dc.identifier.other:1," + col1.getHandle() + ",2"};
```
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: can you also verify a duplicate rowName creates an error?

**Review by benbosman — COMMENTED** [review]
Thanks for this functionality.

I've included some inline comments, but all are rather small requests

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 6]
```diff
@@ -1537,4 +1586,172 @@ public static void main(String[] argv) {
...
+                        newLine.add(key, uuid);
+                    }
+                }
+            } else {
+                if (line.get(key).size() > 1) {
+                    for (String value : line.get(key)) {
+                        newLine.add(key, value);
+                    }
+                } else {
+                    newLine.add(key, line.get(key).get(0));
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: I'm getting this error:
```
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
        at java.util.ArrayList.rangeCheck(ArrayList.java:657)
        at java.util.ArrayList.get(ArrayList.java:433)
        at org.dspace.app.bulkedit.MetadataImport.resolveEntityRefs(MetadataImport.java:1613)
        at org.dspace.app.bulkedit.MetadataImport.runImport(MetadataImport.java:190)
        at org.dspace.app.bulkedit.MetadataImport.main(MetadataImport.java:1526)
```

When doing some debugging I found that:
`line.get(key)` has `[]` and `line.get(key).size()` is `0`
the key is `orgunit.identifier.vat`

and the line:
`{collection=[10400.1/4], dc.date.accessioned=[2019-04-12T13:35:18Z], dc.date.available=[2019-04-12T13:35:18Z], dc.description.provenance[en]=[Made available in DSpace on 2019-04-12T13:35:18Z (GMT). No. of bitstreams: 0], dc.identifier.uri=[http://dev5.rcaap.pt/handle/10400.1/23], orgunit.identifier.city=[Portimão], orgunit.identifier.country=[PT], orgunit.identifier.id=[403925156], orgunit.identifier.name=[Universidade do Algarve - Campus de Portimão], orgunit.identifier.ringgold=[112063], orgunit.identifier.state=[Faro], orgunit.identifier.vat=[], orgunit.identifier.zipcode=[8501-859], relation.isParentOrgUnit=[rowName:76c5cfd9-16ea-4f14-9c9c-b208da26fb29], relationship.type=[OrgUnit], rowName=[b5f709bd-ba08-47e7-9db6-b8e271ad67e0]}`

File: dspace-api/src/main/java/org/dspace/content/dao/impl/MetadataValueDAOImpl.java (hunk 1 above)
Comment (cwilper) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/MetadataValueDAOImpl.java]: DB is not fully optimized for these, but limiting to DSO type (Items) should indeed improve performance. If this were performance-critical, I would advocate for possibly creating an index that involved the text_value, but that would be a rather high cost (space, time) so it would be important to show a large benefit. We will ensure the query *is* limited by DSO type. I also think we could stand to try this on a large test repository and see how ingest rate is affected by value-based vs rowName-based lookups.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java (hunk 2 above)
Comment (cwilper) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Yes, we will add an arg to assertRelationship and update tests to confirm correct order (place) is respected.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java (hunk 3 above)
Comment (cwilper) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Yes it should, and we will add a test case for mixed ref types

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java (hunk 5 above)
Comment (cwilper) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Yes, we will add a test case for two rows declaring the same rowName; this should be an early-detectable error.

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 6 above)
Comment (cwilper) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: @paulo-graca good catch -- we'll fix the issue and add a test case for a csv that contains empty/blank values

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 7]
```diff
@@ -475,7 +500,7 @@ protected void compare(Item item, String[] fromCSV, boolean change,
                                        "item_id=" + item.getID() + ",fromCSV=" + all));
 
         // Don't compare collections  or actions
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: This comment is outdated

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 8]
```diff
@@ -688,6 +714,28 @@ private void handleRelationMetadata(Context c, Item item, String schema, String
...
     }
 
+    /**
+     * Gets an existing entity from a reference.
+     *
+     * @param context the context to use.
+     * @return the entity, which is guaranteed to exist.
+     * @throws MetadataImportException if the reference is badly formed or refers to a non-existing item.
+     */
+    private Entity getEntity(Context context, String reference) throws MetadataImportException {
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Can you include the reference in the JavaDoc, explaining what it expects (UUID or rowNum, …) similar to resolveEntityRef

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 9]
```diff
@@ -1537,4 +1586,172 @@ public static void main(String[] argv) {
...
+        // value reference
+        UUID uuid = null;
+        if (!reference.contains(":")) {
+            // assume it's a UUID
+            try {
+                return UUID.fromString(reference);
+            } catch (IllegalArgumentException e) {
+                throw new MetadataImportException("Not a UUID or indirect entity reference: '" + reference + "'");
+            }
+        } else if (!reference.startsWith("rowName:") ) {
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Can you detail how rowName is handled?

**Review by benbosman — COMMENTED** [review]
I've reviewed all the code now, but haven't tested it yet.
I've only found a few minor things which could use some documentation changes

Since I noticed Paulo has reported a problem while testing the functionality, I will wait with testing until that has been resolved

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 9 above)
Comment (AndrewZWood) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Added some comments in latest commit that should help clarify the fall through to rowName processing.

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 8 above)
Comment (AndrewZWood) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Improved related references docs and added to missing section.

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 7 above)
Comment (AndrewZWood) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Updated.

**Comment by AndrewZWood** [comment]
Pushed up some updates based on feedback thanks everyone. Ready for another round of testing.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+    }
+
+    /**
+     * Test failure when refering to an item in the CSV that hasn't been created yet due to it's order in the CSV
+     */
+    @Test
+    public void testCSVImportWrongOrder() throws Exception {
+        String[] csv = {"id,relationship.type,relation.isAuthorOfPublication,collection,dc.identifier.other",
+                "+,Publication,dc.identifier.other:8675309," + col1.getHandle() + ",2",
+                "+,Person,," + col1.getHandle() + ",1",};
```
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Shouldn't this use 8675309 in the dc.identifier.other column?

**Review by benbosman — COMMENTED** [review]
It seems my feedback has been processed, only one minor comment still.

I will start testing the functionality with actual CSVs next

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 6 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: While testing the new codebase, the error still occurred.
I think the check should be `if (!line.get(key).isEmpty()) {`

**Review by benbosman — CHANGES_REQUESTED** [review]
I've also tested the functionality with a CSV file.

I've noticed 2 things:
* https://github.com/DSpace/DSpace/pull/2471#discussion_r315275012
* Creating a relationship which is not authorized according to the model (e.g. in the wrong direction) will display it would succeed in the preview, but just doesn't create that relationship

**Review by tdonohue — COMMENTED** [review]
@cwilper : Gave this a review today.  The code overall is looking decent, but there's a lot of classes missing JavaDocs (all the new dspace-spring-rest classes have no Javadocs, I just referenced the first few).  There's also some areas of the code that I found very confusing to review (the logic needs more inline comments or enhanced Javadocs, cause I couldn't follow it).  More comments inline below.

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 11]
```diff
@@ -1537,4 +1587,175 @@ public static void main(String[] argv) {
...
+     * Gets a copy of the given csv line with all entity references resolved to UUID strings.
+     *
+     * @param line the csv line to process.
+     * @return a copy, with all references resolved.
+     * @throws MetadataImportException if there is an error resolving any entity reference.
+     */
+    public DSpaceCSVLine resolveEntityRefs(DSpaceCSVLine line) throws MetadataImportException {
+        DSpaceCSVLine newLine = new DSpaceCSVLine(line.getID());
+        for (String key : line.keys()) {
+            if (key.split("\\.")[0].equalsIgnoreCase("relation")) {
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Thanks for the enhanced inline comments here! Much improved. Marking this issue as "resolved"

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 12]
```diff
@@ -1537,4 +1587,175 @@ public static void main(String[] argv) {
...
+        // value reference
+        UUID uuid = null;
+        if (!reference.contains(":")) {
+            // assume it's a UUID
+            try {
+                return UUID.fromString(reference);
+            } catch (IllegalArgumentException e) {
+                throw new MetadataImportException("Not a UUID or indirect entity reference: '" + reference + "'");
+            }
+        } else if (!reference.startsWith("rowName:") ) { //Assume it's a metadata value reference
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: @AndrewZWood : This is much improved too.  Thanks for the detailed explanation and inline comments.  My only small comment here is that the error message could provide a bit more information.  Maybe something like:

`"No matches found for reference: " + reference + ".  Keep in mind you can only reference entries that are listed before this one within the CSV."`

The reason why the second sentence may be important is simply because I can anticipate users asking (on mailing lists etc) why the parser here isn't finding a reference that exists in the CSV (as it is possible that there is a matching entry here, but it's just listed *after* the reference).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+    /**
+     * Test existence of newly created item with proper relationships defined in the item's metadata via
+     * a rowName reference
+     */
+    @Test
+    public void testSingleRowNameRef() throws Exception {
+        String[] csv = {"id,dc.title,relationship.type,relation.isAuthorOfPublication,collection,rowName," +
+                "dc.identifier.other",
+                "+,Test Item 1,Person,," +  col1.getHandle() + ",idVal,0",
+                "+,Test Item 2,Publication,rowName:idVal," +  col1.getHandle() + ",anything,1"};
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: Thanks! Marking resolved

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 6 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Thanks, this works correctly now

**Review by benbosman — COMMENTED** [review]
I've verified that https://github.com/DSpace/DSpace/pull/2471#discussion_r315275012 has been solved

Creating a relationship which is not authorized according to the model (e.g. in the wrong direction) still displays it would succeed in the preview, but just doesn't create that relationship.
I noticed this is actually not an error from this PR, but was already not handled correctly.
It would be great if it can be solved here, otherwise we'd need to resolve this as a separate PR.

The handleRelationOtherMetadata method is the one which handles this use case. It now mentions "no relationship_types were found". The preferred solution would be:
* this is also verified during the preview
* if acceptableRelationshipTypes is not 1, it should block the import

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (AndrewZWood) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Good point, I've gone ahead and added that improvement here: https://github.com/DSpace/DSpace/pull/2471/commits/c4626c1a4940e202e2062989f5eadbd7c9cfa939

**Comment by cwilper** [comment]
@benbosman I wanted to note that @AndrewZWood and I discussed the suggestion about improving validation for the relationship type during review. This makes a lot of sense to do. In the interest of keeping this PR focused and moving it forward, I added this to the list covered by https://jira.duraspace.org/browse/DS-4226, which can be addressed separately (there's already a validation-related improvement listed in there).

**Comment by AndrewZWood** [comment]
Thanks for the additional round of feedback everyone. I believe all outstanding concerns have been addressed.

**Comment by paulo-graca** [comment]
Thank you @cwilper I managed to import my test list! 👍 
I've just 2 questions:
. The import list requires the items to be reordered, you cannot relate items that aren't available at a certain point. Do you know if the export mechanism will be ready to export reordered data (based on the relationships)? Perhaps it will be better to define this as a requirement.
. I did a couple of imports from the same CSV. On all of those, the "relation" was always suggesting to be added. Shouln't it be ignored if it was already present?

```
 + Added   (organization.address.addressCountry): PT
 + Added   (organization.legalName): Universidade do Algarve Escola Superior de Saúde
 + Added   (relation.isParentOrgUnit): 76c5cfd9-16ea-4f14-9c9c-b208da26fb29

```

**Comment by cwilper** [comment]
@paulo-graca glad to hear your test list could be imported with this, thanks for trying it out.

Regarding export, order should not matter because the relationship references in exports will be UUIDs of items that already exist in DSpace. Order only matters on import if you're referring to an item that doesn't yet exist in DSpace.

Your test of importing from the same CSV was helpful. It revealed the fact that we are always treating these relationships as adds, when they should be, like metadata values, treated as "replaces". In other words, if you previously had a relationship and you specify that same relationship, there should be no change as you suggest. In addition, if you specify no relationships, it should be deleted, and if you specify additional relationships, those should be added.

We are looking at code now to see if we can remedy this (with appropriate tests) in this PR. I would also be happy to do this as a followup PR, but if we can fit it in here without too much extra code, as I suspect, it makes sense to bundle with this changeset.

File: .gitignore [hunk 14]
```diff
@@ -42,3 +42,5 @@ nb-configuration.xml
 
 ##Ignore JRebel project configuration
 rebel.xml
+
+test.csv
```
Comment (tdonohue) [inline .gitignore]: Please remove this change to `.gitignore`.  The bug that resulted in a `test.csv` sitting around (after tests) was resolved in #2497.  If you update this PR to latest `master` (via a rebase) you no longer should see the `test.csv`.

File: dspace-api/pom.xml [hunk 15]
```diff
@@ -772,6 +772,12 @@
             <version>2.6.2</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.javax.persistence</groupId>
+            <artifactId>hibernate-jpa-2.1-api</artifactId>
+            <version>1.0.0.Final</version>
+            <scope>compile</scope>
```
Comment (tdonohue) [inline dspace-api/pom.xml]: please remove the `<scope>`, as it's not needed.  "compile" is the default, and all our other dependencies leave this scope out.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+
+    @Autowired
+    private ItemService itemService;
+
+    /**
+     * Setup testing enviorment
+     */
+    @Before
+    public void setup() {
+        context.turnOffAuthorisationSystem();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: I just realized we are turning *off* the authorizationSystem here without ever turning it back on.  This could have unintended consequences in these tests.  We need to turn it back *on* after creating the content in this `setup()` method.

**Review by tdonohue — COMMENTED** [review]
@AndrewZWood : I gave this another code review today.  It's looking mostly great, but I noticed three minor things that would be good to fix prior to merging.  After these three things are fixed, I'm a +1.

**Comment by paulo-graca** [comment]
> Order only matters on import if you're referring to an item that doesn't yet exist in DSpace.

Sorry @cwilper but I've tested this and with an unordered exported list (based on the file I've just send you), I'm getting:
`Error: No matches found for reference: rowName:b5f709bd-ba08-47e7-9db6-b8e271ad67e0,  Keep in mind you can only reference entries that are listed before this one within the CSV.`

**Comment by cwilper** [comment]
> > Order only matters on import if you're referring to an item that doesn't yet exist in DSpace.
> 
> Sorry @cwilper but I've tested this and with an unordered exported list (based on the file I've just send you), I'm getting:
> `Error: No matches found for reference: rowName:b5f709bd-ba08-47e7-9db6-b8e271ad67e0, Keep in mind you can only reference entries that are listed before this one within the CSV`

Thanks for the file @paulo-graca . I immediately noticed you're using rowName:UUID references here and have a duplicate rowName column that is a copy of the id column, which also contains the UUID.

If an item already exists in DSpace, you can just directly use the UUID as the value of the relationship, and skip the rowName stuff. The intent of rowName is to allow you to declare a simple name or unique id within the CSV, to refer to something that *hasn't yet* been ingested. In the case where you use rowName, we do assume that you're referring to an item that is not yet in DSpace, and therefore, order does matter.

But I'm curious about the use case here...if you're pointing out a scenario you think would be useful for some reason I want to make sure that's covered....it may require pre-parsing the CSV and adds a little complexity, but should be doable.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java (hunk 16 above)
Comment (AndrewZWood) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: > @AndrewZWood : I gave this another code review today. It's looking mostly great, but I noticed three minor things that would be good to fix prior to merging. After these three things are fixed, I'm a +1.

Addressed all comments you brought up here: https://github.com/DSpace/DSpace/pull/2471/commits/591df53e7a19e1227e4b0d9db54f9a5bd69db27f#diff-5ce04a4fb4f137714321ddf616041328

**Comment by cwilper** [comment]
@paulo-graca After our discussion on Slack, we were able to reproduce the issue you saw. This is a pre-existing issue we believe should be solved as part of DS-4226 (which we'll be doing a separate PR for).

As discussed, there was actually an error on your initial import, due to a validation step failing. All errors should result in the import appearing to fail, but this one appeared to succeed, so that aspect needs to be fixed. In addition, the reason the validation step failed was that the validation it's doing has a bug: it gets confused when dealing with importing items whose entity type is used on the left side of two separate kinds of relationship types. @AndrewZWood has a better write up of the issue here: https://jira.duraspace.org/browse/DS-4226?focusedCommentId=63162&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-63162

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+                .build();
+        col1 = CollectionBuilder.createCollection(context, parentCommunity).withName("Collection 1").build();
+    }
+
+    /**
+     * Close testing enviorment and restore auth system
+     */
+    @After
+    public void close() {
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: @AndrewZWood : Just a clarification question. Is there a reason why we need authorization turned *off* while running *all tests*?  To me, it looks like our tests are all being kicked off via the commandline (`runDSpaceScript()` method), which means that we should be creating a new Context for each test anyways.  This implies to me that we might be able to do the `restoreAuthSystemState();` *at the very end of the `setup()` method above*.  In other words, it seems like we only need authorization turned off to create the Community/Collection in the `setup()`.  Am I missing something though?

**Review by tdonohue — APPROVED** [review]
👍 Overall, the code now looks great to me. Thanks @AndrewZWood for your hard work on this.  That said, I did ask a small clarification question inline below, that I'd appreciate a response to.  I think we might be able to slightly improve our tests by turning on authorization again *earlier*, but it's also possible I'm overlooking something.

**Review by benbosman — APPROVED** [review]
Thanks @AndrewZWood and @cwilper for this PR.
All issues I had with this PR have been dealt with. This looks good to merge

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java (hunk 17 above)
Comment (AndrewZWood) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CSVMetadataImportReferenceTest.java]: This is a good point, I'll be refactoring `CSVMetadataImportReferenceTest.java` as a part of [DS-4226](https://jira.duraspace.org/browse/DS-4226) and I've noted to verify context authorization requirements at a per test basis.
